### PR TITLE
DAOS-9952 control: Usability improvements to dmg system leader-query

### PR DESCRIPTION
**Pull request by ravalsam**
Adding option to show replicas which are not available (off).

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/cmd/dmg/system.go [hunk 1]
```diff
@@ -69,8 +70,12 @@ func (cmd *leaderQueryCmd) Execute(_ []string) (errOut error) {
 		return cmd.outputJSON(resp, err)
 	}
 
-	cmd.Infof("Current Leader: %s\n   Replica Set: %s\n", resp.Leader,
-		strings.Join(resp.Replicas, ", "))
+	if cmd.ReplicasOff {
+		cmd.Infof("Replicas Off: %s\n", strings.Join(resp.ReplicasOff, ", "))
+	} else {
+		cmd.Infof("Current Leader: %s\n   Replica Set: %s\n  Replicas Off: %s\n", resp.Leader,
+			strings.Join(resp.Replicas, ", "), strings.Join(resp.ReplicasOff, ", "))
```
Comment (mjmac) [inline src/control/cmd/dmg/system.go:77]: minor: I think "Unresponsive Replicas" might be clearer here.

File: src/control/common/proto/ctl/network.pb.go [hunk 2]
```diff
@@ -6,8 +6,8 @@
 
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.26.0
-// 	protoc        v3.6.1
+// 	protoc-gen-go v1.28.1
+// 	protoc        v3.5.0
```
Comment (mjmac) [inline src/control/common/proto/ctl/network.pb.go:10]: Please revert these changes in your tree and re-push. This is what I mentioned in the call the other day -- we don't commit these changes that are only for version updates because they cause a lot of review spam.

**Review by mjmac — CHANGES_REQUESTED** [review]
Generally OK, some minor suggestions for naming. Please revert the version-only changes to the generated pb files and re-push, thanks.

File: src/control/cmd/dmg/system.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ type leaderQueryCmd struct {
 	cfgCmd
 	ctlInvokerCmd
 	jsonOutputCmd
+	ReplicasOff bool `short:"N" long:"off-replica" description:"Show not available Replicas only"`
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/control/cmd/dmg/system.go:49]: Will update

File: src/control/cmd/dmg/system.go (hunk 1 above)
Comment (ravalsam) [reply src/control/cmd/dmg/system.go:77]: Will update

File: src/control/cmd/dmg/system.go [hunk 4]
```diff
@@ -69,8 +70,12 @@ func (cmd *leaderQueryCmd) Execute(_ []string) (errOut error) {
 		return cmd.outputJSON(resp, err)
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/control/cmd/dmg/system.go:70]: Well, it provided the full output for both the option. Do we want to alter and only show DownReplicas for json?

```shell
[samirrav@wolf-170 EC_Demo]$ dmg -o /home/samirrav/scripts/EC_Demo/test_daos_dmg.yaml system leader-query -j
{
  "response": {
    "CurrentLeader": "10.8.1.178:10001",
    "Replicas": [
      "10.8.1.173:10001",
      "10.8.1.174:10001",
      "10.8.1.175:10001",
      "10.8.1.176:10001",
      "10.8.1.178:10001"
    ],
    "DownReplicas": [
      "10.8.1.173:10001"
    ]
  },
  "error": null,
  "status": 0
}
[samirrav@wolf-170 EC_Demo]$ dmg -o /home/samirrav/scripts/EC_Demo/test_daos_dmg.yaml system leader-query -N -j
{
  "response": {
    "CurrentLeader": "10.8.1.178:10001",
    "Replicas": [
      "10.8.1.173:10001",
      "10.8.1.174:10001",
      "10.8.1.175:10001",
      "10.8.1.176:10001",
      "10.8.1.178:10001"
    ],
    "DownReplicas": [
      "10.8.1.173:10001"
    ]
  },
  "error": null,
  "status": 0
}

```

File: src/control/lib/control/system.go [hunk 5]
```diff
@@ -639,7 +641,24 @@ func LeaderQuery(ctx context.Context, rpcClient UnaryInvoker, req *LeaderQueryRe
 	}
 
 	resp := new(LeaderQueryResp)
-	return resp, convertMSResponse(ur, resp)
+	if err = convertMSResponse(ur, resp); err != nil {
+		return nil, errors.Wrap(err, "converting MS to LeaderQuery resp")
+	}
+
+	req.SetHostList(resp.Replicas)
+	ur, err = rpcClient.InvokeUnaryRPC(ctx, req)
```
Comment (kjacque) [inline src/control/lib/control/system.go:649]: One thing I find myself wondering is whether the MS could find this information on the first call, rather than requiring two RPCs. Shouldn't the MS have that information in its database?

**Review by kjacque — COMMENTED** [review]
Overall looks okay. One question though.

File: src/control/cmd/dmg/system.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/dmg/system.go:70]: I think this is fine as-is. Any JSON consumer is going to pick apart the data programmatically anyhow.

As you're going to need to push an update to revert most of the generated protobuf files, I also suggest that you may want to fix the JSON output for this struct so that it uses `snake_case` field names. We have been trying to standardize on that for all of the structs, but there are a few here and there that haven't been updated. 

You'd just need to update the struct here: https://github.com/daos-stack/daos/blob/master/src/control/lib/control/system.go#L622

The updated field names would be: `current_leader`, `replicas`, `down_replicas`. Looks like you would need to update a couple of tests though (`grep -Irs CurrentLeader src/tests`), so I won't insist on that change as it's expanded scope for this PR.

File: src/control/common/proto/ctl/firmware.pb.go [hunk 6]
```diff
@@ -6,8 +6,8 @@
 
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.26.0
-// 	protoc        v3.6.1
+// 	protoc-gen-go v1.28.1
+// 	protoc        v3.5.0
```
Comment (mjmac) [inline src/control/common/proto/ctl/firmware.pb.go:10]: These version-only changes to generated files should not be part of the commit. You can revert them locally by doing `git checkout master -- path/to/pb.go` for the files that only have version updates, then pushing that commit so that they are not included in the landing commit.

File: src/control/lib/control/system.go (hunk 5 above)
Comment (ravalsam) [reply src/control/lib/control/system.go:649]: Yes, it's store in the database. But we currently store the only host name and not the status of the raft. So if replicas are down, we need to send another RPC to those replicas to make sure all are up and responsive.

**Review by mjmac — CHANGES_REQUESTED** [review]
Sorry, perhaps I wasn't clear enough in my previous review. There are many changes to generated protobuf files that should not be included in the commit because they are not related to the new functionality. As a team, we do not allow landing of these changes because they would result in unnecessary churn when developers with slightly different versions of the tools work on the protobuf files. Please revert the changes that are not related to the new field in `svc.proto`.

File: src/control/lib/control/system.go (hunk 5 above)
Comment (kjacque) [reply src/control/lib/control/system.go:649]: Thanks for the explanation.

File: src/control/common/proto/ctl/network.pb.go (hunk 2 above)
Comment (ravalsam) [reply src/control/common/proto/ctl/network.pb.go:10]: Sorry, I didn't get this initially. I will change all the files from master which has only proto version changes.

**Review by mjmac — APPROVED** [review]

**Review by kjacque — DISMISSED** [review]

**Review by tanabarr — CHANGES_REQUESTED** [review]
There are unrelated proto auto generated file changes still, only `svc.proto` was updated and so only `svc.{pb.go,pb-c.[ch]}` proto files should be updated. Given you have to re-push, may as well update CurrentLeader json tag to currently_leader. Thanks

File: src/control/lib/control/system.go [hunk 7]
```diff
@@ -615,13 +615,15 @@ func SystemErase(ctx context.Context, rpcClient UnaryInvoker, req *SystemEraseRe
...
 	msRequest
+	sysRequest
 }
 
 // LeaderQueryResp contains the status of the request and, if successful, the
 // MS leader and set of replicas in the system.
 type LeaderQueryResp struct {
-	Leader   string `json:"CurrentLeader"`
-	Replicas []string
+	Leader       string   `json:"CurrentLeader"`
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/system.go:624]: ok, no problem that's not a blocking issue

**Comment by ravalsam** [comment]
> There are unrelated proto auto generated file changes still, only `svc.proto` was updated and so only `svc.{pb.go,pb-c.[ch]}` proto files should be updated. Given you have to re-push, may as well update CurrentLeader json tag to currently_leader. Thanks

Done

**Review by kjacque — DISMISSED** [review]

**Review by tanabarr — APPROVED** [review]

**Review by tanabarr — CHANGES_REQUESTED** [review]
You still have changes committed to generated protobuf files in this PR without any relevant .proto definition updates. This needs to be fixed. See smd and storage_nvme auto compiled files.

**Comment by ravalsam** [comment]
> Hello Samir. You still have changes committed to generated protobuf files in this PR without any relevant .proto definition updates. This needs to be fixed. See smd and storage_nvme auto compiled files. The only Proto definition file that was updated is svc.proto therefore the only auto compiled protobuf files that should be committed with changes are svc.pb.go and svc.pb-c.[ch].

Those file changed because of comments also updated in the code and for *proto was not updated. I was asking about the same on Team chat and got the answer that we should also update those files.

**Review by kjacque — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]

**Review by mjmac — APPROVED** [review]
